### PR TITLE
apple envy moe implementation

### PR DESCRIPTION
**Pull request by entrpn**
# Description

Implements Apple Envy MoE model as described [here](https://github.com/apple/axlearn/blob/cbdcde5e745ba29ef84eb82e11f2aaa7af24dca1/axlearn/experiments/text/gpt/envy.py)
# Tests

# Envy MoE Parameter Equivalence Report

This report compares the parameter structures of Apple's Envy Switch-Base model between **AXLearn** and **MaxText** to verify architectural correctness.

## 1. Parameter Shape Mapping

Here is the direct mapping of parameter tensors for the **12-layer Switch-Base model** (6 repeat blocks of Alternating Dense and Sparse MoE layers):

| Component / Layer | AXLearn Parameter (Shape) | MaxText Parameter (Shape) | Equivalence Note |
| :--- | :--- | :--- | :--- |
| **Token Embedding / Logits** | `decoder/emb/token_emb/weight` `[32768, 1536]` | `token_embedder/embedding` `(32768, 1536)` | **Exactly Shared & Identical** (`logits_via_embedding: true`) |
| **Final Decoder Norm** | `decoder/output_norm/scale` `[1536]` | `decoder_norm/scale` `(1536,)` | **Exactly Identical** |
| **Self-Attention Norm** | `self_attention/norm/scale` `(6, 1536)` | `pre_self_attention_layer_norm/scale` `(1536, 6)` | **Exactly Identical** (Transposed layers axis) |
| **Self-Attention QKV Proj**| `self_attention/attention/i_proj/i_proj/qkv_proj/weight` `(6, 1536, 36, 128)` | Q, K, V are stored as 3 separate parameters:<br>- `query/kernel` `(1536, 6, 12, 128)`<br>- `key/kernel` `(1536, 6, 12, 128)`<br>- `value/kernel` `(1536, 6, 12, 128)` | **Exactly Identical**.<br>AXLearn groups QKV along the projection axis:<br>$12 + 12 + 12 = 36$ heads total. |
| **Self-Attention QK Norm** | `attention/scale_query/norm/scale` `(6, 128)` <br> `attention/scale_key/norm/scale` `(6, 128)` | `query_norm/scale` `(128, 6)` <br> `key_norm/scale` `(128, 6)` | **Exactly Identical** (Transposed layers axis) |
| **Self-Attention Output Proj**| `self_attention/attention/o_proj/weight` `(6, 1536, 12, 128)` | `out/kernel` `(12, 6, 128, 1536)` | **Exactly Identical** |
| **Dense FFN Norm** | `feed_forward/prenorm/scale` `(6, 1536)` | `post_self_attention_layer_norm/scale` `(1536, 6)` | **Exactly Identical** (Transposed layers axis) |
| **Dense FFN Post-Norm** | `feed_forward/postnorm/scale` `(6, 1536)` | `post_ffw_norm/scale` `(1536, 6)` | **Exactly Identical** (`use_post_ffw_norm: true`) |
| **Dense FFN Projections** | `feed_forward/linear1_0/weight` `(6, 1536, 6144)` <br> `feed_forward/linear1_1/weight` `(6, 1536, 6144)` <br> `feed_forward/linear2/weight` `(6, 6144, 1536)` | `wi_0/kernel` `(1536, 6, 6144)` <br> `wi_1/kernel` `(1536, 6, 6144)` <br> `wo/kernel` `(6144, 6, 1536)` | **Exactly Identical** |
| **Sparse MoE Norm** | `feed_forward/prenorm/scale` `(6, 1536)` | `post_self_attention_layer_norm/scale` `(1536, 6)` | **Exactly Identical** (Transposed layers axis) |
| **Sparse MoE Post-Norm** | `feed_forward/postnorm/scale` `(6, 1536)` | `post_ffw_norm/scale` `(1536, 6)` | **Exactly Identical** (`use_post_ffw_norm: true`) |
| **Sparse MoE Gate** | `feed_forward/gate_weight` `(6, 1536, 128)` | `MoeBlock_0/gate/kernel` `(1536, 6, 128)` | **Exactly Identical** |
| **Sparse MoE Experts Proj**| `feed_forward/wi_0_weight` `(6, 128, 1536, 6144)` <br> `feed_forward/wi_1_weight` `(6, 128, 1536, 6144)` <br> `feed_forward/wo_weight` `(6, 128, 6144, 1536)` | `MoeBlock_0/wi_0` `(128, 6, 1536, 6144)` <br> `MoeBlock_0/wi_1` `(128, 6, 1536, 6144)` <br> `MoeBlock_0/wo` `(128, 6, 6144, 1536)` | **Exactly Identical** |

---

## 2. Parameter Counts Analysis

- **AXLearn Parameter Count**: `22,077,958,656` (22.07B)
- **MaxText Parameter Count**: `22,077,958,656` (22.07B)
- **Difference**: **`0` parameters (100% exact match)**

### Key Configurations Enabled:
1. **`logits_via_embedding: true`**: Shares the token embedding parameters with the output logits projection.
2. **`use_post_ffw_norm: true`**: Activates the post-FFN normalization layer (`post_ffw_norm` in MaxText, mapping to `postnorm` in AXLearn's `hybridnorm` FFN structure).

Ran test:

```
python -m pytest tests/unit/envy_test.py
=========================================================== test session starts ===========================================================
platform linux -- Python 3.12.11, pytest-8.4.2, pluggy-1.6.0
rootdir: /mnt/disks/external_disk/maxtext
configfile: pytest.ini
plugins: anyio-4.13.0, typeguard-2.13.3, jaxtyping-0.3.9, xdist-3.8.0, hypothesis-6.151.9, mock-3.15.1
collected 1 item                                                                                                                          

tests/unit/envy_test.py .                                                                                                           [100%]

============================================================ warnings summary =============================================================
maxtext_venv/lib/python3.12/site-packages/flax/nnx/variablelib.py:324
  /mnt/disks/external_disk/maxtext/maxtext_venv/lib/python3.12/site-packages/flax/nnx/variablelib.py:324: DeprecationWarning: jax.core.Effect is deprecated. Use jax.extend.core.Effect.
    class VariableEffect(jax.core.Effect): ...

tests/unit/envy_test.py::EnvyTest::test_model_initialization_and_forward
  /mnt/disks/external_disk/maxtext/maxtext_venv/lib/python3.12/site-packages/flax/core/tracers.py:30: DeprecationWarning: jax.core.get_opaque_trace_state is deprecated. Use jax.extend.core.get_opaque_trace_state.
    return jax.core.get_opaque_trace_state(convention="flax")

tests/unit/envy_test.py: 90 warnings
  /mnt/disks/external_disk/maxtext/src/maxtext/layers/nnx_scan.py:84: DeprecationWarning: '.value' access is now deprecated. For Variable[Array] instances use:
  
    variable[...]
  
  For other Variable types use:
  
    variable.get_value()
  
    is_leaf=lambda x: hasattr(x, "replace") and hasattr(x, "value"),

tests/unit/envy_test.py: 90 warnings
  /mnt/disks/external_disk/maxtext/src/maxtext/layers/nnx_scan.py:56: DeprecationWarning: '.value' access is now deprecated. For Variable[Array] instances use:
  
    variable[...]
  
  For other Variable types use:
  
    variable.get_value()
  
    if hasattr(leaf, "replace") and hasattr(leaf, "value"):

tests/unit/envy_test.py: 90 warnings
  /mnt/disks/external_disk/maxtext/src/maxtext/layers/nnx_scan.py:65: DeprecationWarning: 'sharding_names' is deprecated, use 'out_sharding' instead.
    value = getattr(leaf, key, None)

tests/unit/envy_test.py::EnvyTest::test_model_initialization_and_forward
tests/unit/envy_test.py::EnvyTest::test_model_initialization_and_forward
  /mnt/disks/external_disk/maxtext/maxtext_venv/lib/python3.12/site-packages/qwix/_src/qconfig.py:178: DeprecationWarning: The 'm.iter_modules()' method is deprecated; use the 'nnx.iter_modules(m)' function instead.
    for _, node in model.iter_modules():

tests/unit/envy_test.py::EnvyTest::test_model_initialization_and_forward
tests/unit/envy_test.py::EnvyTest::test_model_initialization_and_forward
  /mnt/disks/external_disk/maxtext/maxtext_venv/lib/python3.12/site-packages/flax/nnx/statelib.py:378: DeprecationWarning: `flax.nnx.State` will be deprecated and be replaced by the built-in Python dict. Please use the equivalent `nnx.split_state` instead.
    warnings.warn(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
==================================================== 1 passed, 276 warnings in 14.50s =====================================================
sys:1: DeprecationWarning: builtin type swigvarlink has no __module__ attribute
```

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/configs/models/envy-switch-base.yml [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+base_num_decoder_layers: 12
+base_num_query_heads: 12
+base_num_kv_heads: 12
+base_mlp_dim: 6144
+base_moe_mlp_dim: 6144
+vocab_size: 32768
+normalization_layer_epsilon: 1e-05
+rope_max_timescale: 500000
+num_experts: 128
+num_experts_per_tok: 1
```
Comment (JamesDeng42) [inline src/maxtext/configs/models/envy-switch-base.yml:33]: the apple's repo uses top2 routing instead of 1?

File: src/maxtext/configs/models/envy-switch-large.yml [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Model configuration for Envy Switch-Large (104B)
+
+decoder_block: "envy"
+mlp_activations: ["silu", "linear"]
+enable_dropout: false
+logits_via_embedding: true
+tokenizer_type: "huggingface"
```
Comment (JamesDeng42) [inline src/maxtext/configs/models/envy-switch-large.yml:21]: we also need to set the load balance loss to 0.01 since apple's repo use load balance loss by default ?

**Review by JamesDeng42 — APPROVED** [review]
left some comments, otherwise, it looks good to me.

File: src/maxtext/configs/models/envy-switch-base.yml (hunk 1 above)
Comment (entrpn) [reply src/maxtext/configs/models/envy-switch-base.yml:33]: The default value for TransformerFeedForwardMoE which is what envy is using in axlearn is 1: https://github.com/apple/axlearn/blob/main/axlearn/common/mixture_of_experts.py#L238

This also is what the switch transformer uses which is the basis of this model.

File: src/maxtext/configs/models/envy-switch-large.yml (hunk 2 above)
Comment (entrpn) [reply src/maxtext/configs/models/envy-switch-large.yml:21]: added.

File: src/maxtext/configs/models/envy-switch-base.yml [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+capacity_factor: 2.0
```
Comment (RissyRan) [inline src/maxtext/configs/models/envy-switch-base.yml]: For capacity_factor is a implementation details (dropping/dropless), not related to model config itself.

Recommend to remove from here. If you'd like to assert this always happen, you could add an assertion in types.py file and ensure envy code block is called with capacity_factor=2

File: src/maxtext/configs/models/envy-switch-base.yml [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+use_qk_norm: true
+use_post_ffw_norm: true
+load_balance_loss_weight: 0.01
+interleave_moe_layer_step: 2
+inhomogeneous_layer_cycle_interval: 2
+max_target_length: 8192
```
Comment (RissyRan) [inline src/maxtext/configs/models/envy-switch-base.yml]: Similar comment with capacity factor.

File: src/maxtext/configs/models/envy-switch-base.yml (hunk 3 above)
Comment (RissyRan) [reply src/maxtext/configs/models/envy-switch-base.yml]: Similar comments for other envy configs.

File: tests/unit/envy_test.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+from maxtext.configs import pyconfig
+from maxtext.common.common_types import MODEL_MODE_TRAIN
+from maxtext.layers import quantizations
+from maxtext.models import models
+from maxtext.utils import maxtext_utils
+from tests.utils.test_helpers import get_test_config_path
+import numpy as np
+
+
+class EnvyTest(unittest.TestCase):
```
Comment (RissyRan) [inline tests/unit/envy_test.py:32]: Does this test functionality of this model? If so, you could write a compile test like this: https://github.com/AI-Hypercomputer/maxtext/blob/025f83f7174f74705eb3b0e76548f57637121015/tests/unit/train_compile_test.py#L866

**Review by RissyRan — COMMENTED** [review]
LGTM at high level. Will you test logits alignment in next PR for the model?

File: src/maxtext/layers/decoders.py [hunk 6]
```diff
@@ -1078,6 +1082,11 @@ def __call__(
                 "nope_layer_interval": self.config.nope_layer_interval,
                 "interleave_moe_layer_step": self.config.interleave_moe_layer_step,
             }
+          if cfg.decoder_block == DecoderBlockType.ENVY:
```
[the remark this answers is not in the document]
Comment (entrpn) [reply src/maxtext/layers/decoders.py:1085]: I'm going to keep it as is as it follows the current maxtext convention

File: src/maxtext/layers/decoders.py [hunk 7]
```diff
@@ -1203,6 +1212,11 @@ def __call__(
                   "is_nope_layer": llama4.determine_is_nope_layer(lyr, self.config.nope_layer_interval),
                   "is_moe_layer": llama4.determine_is_moe_layer(lyr, self.config.interleave_moe_layer_step),
               }
+            if cfg.decoder_block == DecoderBlockType.ENVY:
```
[the remark this answers is not in the document]
Comment (entrpn) [reply src/maxtext/layers/decoders.py:1215]: same as above, will keep it to keep the same convention that is currently used in maxtext.

File: src/maxtext/configs/models/envy-switch-base.yml (hunk 3 above)
Comment (entrpn) [reply src/maxtext/configs/models/envy-switch-base.yml]: will remove capacity_factor from config and do a check in types.py for now to match the axlearn implementation.

File: src/maxtext/configs/models/envy-switch-base.yml (hunk 4 above)
Comment (entrpn) [reply src/maxtext/configs/models/envy-switch-base.yml]: same as above, will fix it.

Comment (RissyRan) [reply src/maxtext/configs/models/envy-switch-base.yml]: It seems still there?

File: src/maxtext/configs/models/envy-switch-large.yml [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+normalization_layer_epsilon: 1e-05
+rope_max_timescale: 500000
+num_experts: 128
+num_experts_per_tok: 1
+use_qk_norm: true
+use_post_ffw_norm: true
+load_balance_loss_weight: 0.01
+interleave_moe_layer_step: 2
+inhomogeneous_layer_cycle_interval: 2
+max_target_length: 8192
```
Comment (RissyRan) [inline src/maxtext/configs/models/envy-switch-large.yml]: Same comment

File: src/maxtext/configs/models/envy-switch-xxl.yml [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+normalization_layer_epsilon: 1e-05
+rope_max_timescale: 500000
+num_experts: 64
+num_experts_per_tok: 1
+use_qk_norm: true
+use_post_ffw_norm: true
+load_balance_loss_weight: 0.01
+interleave_moe_layer_step: 2
+inhomogeneous_layer_cycle_interval: 2
+max_target_length: 8192
```
Comment (RissyRan) [inline src/maxtext/configs/models/envy-switch-xxl.yml]: Same comment

File: tests/unit/envy_test.py (hunk 5 above)
Comment (RissyRan) [reply tests/unit/envy_test.py:32]: If same functionality, you could optionally remove this test (to save some CI tests). But it's up to you.

**Review by RissyRan — COMMENTED** [review]
LGTM! Are you planning to test logits alignment?

**Review by RissyRan — APPROVED** [review]

**Comment by entrpn** [comment]
Closing and using different PR: https://github.com/AI-Hypercomputer/maxtext/pull/4695
